### ISSUE TITLE
I'm cleaning up the docs, and have some questions

### ISSUE THREAD

**Issue by Pyrohh** [neovim/neovim#2231]
I'm trying to do a pass over the docs to see what things can be removed, and have a few questions:

- Do we need any of the `if_[language].txt` docs (besides python)? If we're not going to support most language specific plugin interfaces then we should probably remove their respective docs (which can be easily reverted if we do support them).
- What needs to be preserved in the `gui_*.txt` docs? I'd like to unify them if possible, and a lot of the stuff in there is outdated (pre-X11 references everywhere).
- Do we need `pi_(vimball|getscript).txt`? I'm curious if they're even used anymore.
- `remote.txt`? After reading #1750, it sounds like msgpack offers better options, but nothing has been implemented as of yet
- @tarruda do you know how much of `term.txt` is still relevant? I'm guessing unibilium abstracts a lot away.

**Comment by elmart** [comment]
Some time ago, while doing something else, I noticed there were many hidden options (see `:help hidden-options`) that weren't correctly marked as such in the docs. They are out-of-sync to this respect. 

That's easy to update. Options array is located at `option.c:384`. This is:
```
static vimoption_T
  options[] =
{
  {"aleph",       "al",   P_NUM|P_VI_DEF|P_CURSWANT,
   (char_u *)&p_aleph, PV_NONE,
   {
     (char_u *)224L,
     (char_u *)0L
   } SCRIPTID_INIT},
  {"antialias",   "anti", P_BOOL|P_VI_DEF|P_VIM|P_RCLR,
   (char_u *)NULL, PV_NONE,
   {(char_u *)FALSE, (char_u *)FALSE}
   SCRIPTID_INIT},
...
```
Hidden options have "var" field (4th one) set to NULL. In the above, you can see `antialias` is a hidden option. But it's not marked as that in the docs. There are many like that.

I had this task in my todo list. But, if you want to take it, I'll be glad.

**Comment by tarruda** [comment]
> Do we need any of the if_[language].txt docs (besides python)? If we're not going to support most language specific plugin interfaces then we should probably remove their respective docs (which can be easily reverted if we do support them).

:+1: 

> What needs to be preserved in the gui_*.txt docs? I'd like to unify them if possible, and a lot of the stuff in there is outdated (pre-X11 references everywhere).

I think that can all be safely removed

> Do we need pi_(vimball|getscript).txt? I'm curious if they're even used anymore.

I didn't even knew about these. IMO these should be removed as its rare seeing someone not using a package manager these days. cc @justinmk 

> remote.txt? After reading #1750, it sounds like msgpack offers better options, but nothing has been implemented as of yet

Lets leave it for now, I plan reimplement those flags on top of msgpack-rpc

> @tarruda do you know how much of term.txt is still relevant? I'm guessing unibilium abstracts a lot away.

Most of it seems to be irrelevant(all terminal options are gone, for example). I think it can be safely removed.

**Comment by tarruda** [comment]
Thanks for doing this @Pyrohh :+1:

**Comment by Pyrohh** [comment]
@elmart Thanks, but I have enough to do for now. If I added that to my already long list it would never get done :)

@tarruda Thanks for the suggestions, I'll wait for some more opinions before starting.

**Comment by Pyrohh** [comment]
@tarruda Is raw mode still of importance, or is it sufficiently abstracted away by the new terminal libraries to be mentioned at all? I ask because these two comments explicitly mention it, and I'm unsure if they still have any relevance:

https://github.com/neovim/neovim/blob/master/src/nvim/main.c#L363
https://github.com/neovim/neovim/blob/master/src/nvim/fileio.c#L1802

**Comment by justinmk** [comment]
- What needs to be preserved in the `gui_*.txt` docs? I'd like to unify them if possible, and a lot of the stuff in there is outdated (pre-X11 references everywhere).

- `x11-selection` section in `gui_x11.txt`. Should be merged into `:h clipboard`, with only the essential parts. Our goal should be to minimize platform-specific nuances (in documentation and implementation :)
- `gui_win32.txt` has at least 2 sections that warrant a closer look, but I don't want to review it right now. I suggest keeping it for now. The reason is: yes, we can always pull deleted things from source control, but the different between *not deleting it* and *deleting it* is that in the former case, we have a living reminder of what remains to be dealt-with, and in the latter case, we have no idea where to start when the time comes to document Win32 behavior.

- Do we need `pi_(vimball|getscript).txt`? I'm curious if they're even used anymore.

We should not support vimball/getscript at all. :cold_sweat: <-- exhausted from resisting snarky comments about vimball.

- @tarruda do you know how much of `term.txt` is still relevant? I'm guessing unibilium abstracts a lot away.

`:h slow-terminal` still has relevant passages, e.g.:

> If you have a slow terminal you may want to reset the 'showcmd' option.

Let's keep that section (or move it) and figure out which parts are no longer relevant, and explain clearly (in this thread, at least) the reason why some options like `ttimeout` behave differently or are not relevant. That saves us the hassle of uncovering these questions one-by-one as confused users find their way here.

Most of `:h mouse-using` seems relevant to Neovim as well.

So everything but sections 4 & 5 can be removed from term.txt.

**Comment by justinmk** [comment]
>  I noticed there were many hidden options (see :help hidden-options) that weren't correctly marked as such in the docs

In addition to that, it has always bugged me that `v:` variables are not consistently documented as read-only. In fact we should only mark the ones that are *writable* since there should be few. And there should be a quick-reference table at `:help v:`.

**Comment by chrisbra** [comment]
Hi Michael!

On Mo, 23 Mär 2015, Michael Reed wrote:

> - Do we need `pi_(vimball|getscript).txt`? I'm curious if they're even used anymore.

Please keep the vimball feature it is generally useful if one can't or 
don't want to use a plugin manager. This is one of the underestimated 
features of Vim that works really well in practice and a reason, why I 
still distribute my plugins as vimballs.

Best,
Christian
-- 
Frankreich ist ein unglaublich fruchbares Land.
Man sät ein paar Staatsbeamte und erntet wenig später neue Steuern.
		-- Edmond de Goncourt

**Comment by justinmk** [comment]
Sigh.

**Comment by tarruda** [comment]
@chrisbra The only reason I see for not using a plugin manager is to install plugins in a machine without git or internet connection. Even in this case, I doubt anyone would want to install each plugin individually, so there is a better alternatives such as a tarball with the `.vim` directory or  [myvim](https://github.com/junegunn/myvim) (kudos to @junegunn for making such great tools/plugins)

This is just my opinion, of course. Can you imagine any other scenario where installing with vimball would be useful?

**Comment by fwalch** [comment]
FWIW, I don't like the concept of vimballs either (or maybe I just don't understand it well enough), but if people still use it and we don't really gain much by removing it, let's just keep it.
Hope that won't exhaust you too much, @justinmk :grin:

**Comment by chrisbra** [comment]
Am 2015-03-24 09:07, schrieb Thiago de Arruda:
> @chrisbra [1] The only reason I see for not using a plugin manager is
> to install plugins in a machine without git or internet connection.
> Even in this case, I doubt anyone would want to install each plugin
> individually, so there is a better alternatives such as a tarball with
> the .vim directory or myvim [2] (kudos to @junegunn [3] for making
> such great tools/plugins)
> 
> This is just my opinion, of course. Can you imagine any other scenario
> where installing with vimball would be useful?

I have a work station, where neither git is available nor github is 
accessible.

Since the work station is a Windows machine, tar is not an option, plus
even than I would have to generally run :helptags after installing a 
plugin manually.

And even if I had tar installed, I would have to manually select the 
destination
directory. With vimballs I don't have to take care, Vim does everything.

So I generally download vimballs from vim.org click on the downloaded 
files and they
open automatically in Vim then I run :so % and restart Vim. This is more 
comfortable
than anything else and all I need.

This can even be done for several plugins at once, although I usually do 
not download
several plugins manually this way.

Also vimballs are not useful for binary files in which case I have to go 
to the
trouble of installing zipfiles manually, but that is much more 
cumbersome. But
this is not a big problem in practice, as most plugins only contain text 
files.

This is all I need and vimballs works very well this way. I don't have 
the need for
automatic updates, if a plugin works, I keep it, else I will uninstall 
it.
Even on my other personal private work station I don't have a plugin 
manager
installed, as I don't have a need for it.

**Comment by justinmk** [comment]
> but if people still use it and we don't really gain much by removing it, let's just keep it. 

- Saving users the hassle of evaluating poorly-designed options is a good thing. Continuing to make misfeatures available harms usability. This isn't meant as snark, it's just a fact.
- Reducing the surface area of supported options is good for the project.
- Why should we support environments that don't have git or some other pseudo-plugin-manager? Would such an environment be likely to run Neovim?
- How many Neovim users use vimballs? I estimate zero right now. Should we keep features that zero users use?

**Comment by jszakmeister** [comment]
> Saving users the hassle of evaluating poorly-designed options is a good thing. Continuing to make misfeatures available harms usability. This isn't meant as snark, it's just a fact.

I don't see what makes Vimballs a "poorly-designed" option.  In fact, I could argue the other way.  Since there isn't senseless expanding of the runtime path, it's better in that there are less places to look after it's installed.

> Reducing the surface area of supported options is good for the project.

Wasn't the goal to be compatible with Vim?

> Why should we support environments that don't have git or some other pseudo-plugin-manager? Would such an environment be likely to run Neovim?

I've got a fleet of computers that are completely disconnected from the Internet--by design, for security reasons.  That won't change.  There are other organizations that are doing the same too.  I use Vim in that environment *all the time*.  Serious development happens there.  I'd love to run Neovim there, once it's ready to be my daily editor.

> How many Neovim users use vimballs? I estimate zero right now...

When Neovim becomes more stable and reaches a wider audience, I doubt that will remain the case.  OTOH, can't vimball support be added as a plugin?  I thought Dr. Chip maintained it.

**Comment by tarruda** [comment]
> I don't see what makes Vimballs a "poorly-designed" option. In fact, I could argue the other way. Since there isn't senseless expanding of the runtime path, it's better in that there are less places to look after it's installed.

I think what he meant is that its poorly designed for managing individual plugins when compared to a package manager. While vimballs keeps a manifest of installed files, you need one vimball per plugin(imagine a user that wants to install 100+ plugins having to source each vimball individually). BTW Emacs bundles its own package manager and I think we should eventually do the same(vote for vim-plug :D)

**Comment by justinmk** [comment]
> there are less places to look after it's installed.

And it interleaves the plugin's various files with other plugins' files, which means if you update a plugin there could be painful conflicts that the user must sort out, thus losing any time saved by adoption of vimball. Not long ago I tried to update netrw using its vimball, and it left the old one hanging around.

> Wasn't the goal to be compatible with Vim?

Until now I had not heard anyone speak in favor of continuing to ship the entirety of the Vim runtime files with Neovim. Personally, I strongly prefer compatibility in most cases, I just did not think vimball was such a case.

> can't vimball support be added as a plugin?

I suppose in this case it's a bit of a chicken-and-egg problem :)

**Comment by chrisbra** [comment]
>> there are less places to look after it's installed.
> 
> And it interleaves the plugin's various files with other plugins'

That is what the runtimepath was meant for, isn't it?

> files, which means if you update a plugin there could be painful
> conflicts that the user must sort out, thus losing any time saved by
> adoption of vimball.

Usually installing a vimball on top of an already installed plugin 
works.

> Not long ago I tried to update netrw using its
> vimball, and it left the old one hanging around.

IIRC netrw is special, since it is distributed with vim and its 
runtimepath
is usually not writable by a normal user.
But, usually you can simply use :RmVimball and that should work.

BTW: Who needs hundreds of plugins installed all at once?

**Comment by tarruda** [comment]
> Until now I had not heard anyone speak in favor of continuing to ship the entirety of the Vim runtime files with Neovim.

Once I envisioned distributing nvim with a very clean runtime directory(only the bare essentials, such as a package manager). Even language syntax and filetype-specific scripts would be left out. The reason is that people usually will install third-party plugins for programming in a certain language even if vim already has a syntax file for the language, so we might as well avoid the overhead of maintaining these things in the repository.

Assuming we eventually distribute a package manager with Neovim, the following approach could be taken when the file edits a file that contains no syntax script:

```
:e script.py
You have no syntax files for python, would you like to install a plugin for python editing?(Y/n)
```

If the user answers yes, a list of known python plugins is displayed and the plugin is installed/sourced automatically by the plugin manager.

wc -l runtime/**/*.vim shows 179k lines of vimscript, I wonder how much of that is actually used by the regular vim user?

**Comment by tarruda** [comment]
> BTW: Who needs hundreds of plugins installed all at once?

I don't have a hundred, but my vimrc has about 50 plugins, and when I ssh into a new machine the only thing I have to do is invoke vim-plug to mirror my local nvim installation

**Comment by chrisbra** [comment]
> :e script.py
> You have no syntax files for python, would you like to install a
> plugin for python editing?(Y/n)
> 
> If the user answers yes, a list of known python plugins is displayed
> and the plugin is installed/sourced automatically by the plugin
> manager.

And if one doesn't have a working network connection?

> wc -l runtime/*_/_.vim shows 179k lines of vimscript, I wonder how
> much of that is actually used by the regular vim user?

Depends on the definition of regular user. But does that hurt to 
distribute it?

**Comment by jszakmeister** [comment]
> I don't have a hundred, but my vimrc has about 50 plugins, and when I ssh into a new machine the only thing I have to do is invoke vim-plug to mirror my local nvim installation

But you don't necessarily need to install them all at once.  For quite a while, I used nothing but vimballs and versioned my whole vimfiles area.  I use Pathogen now.  That decision revolves more around multiple people maintaining a core set of vimfiles, than anything else though (the vimball manifest didn't translate well across multiple users and trying to upgrade plugins--it worked well for a single user though).

**Comment by tarruda** [comment]
> And if one doesn't have a working network connection?

Assuming the user wants to move his setup to a server or machine without internet, something like [myvim](https://github.com/junegunn/myvim) works nicely.

> Depends on the definition of regular user. But does that hurt to
distribute it?

No, but it bothers me having this amount of code in the repository when I know most of it isn't used and has been abandoned for years.

**Comment by jszakmeister** [comment]
> Once I envisioned distributing nvim with a very clean runtime directory(only the bare essentials, such as a package manager). Even language syntax and filetype-specific scripts would be left out. The reason is that people usually will install third-party plugins for programming in a certain language even if vim already has a syntax file for the language, so we might as well avoid the overhead of maintaining these things in the repository.

I don't have any issues about reducing what we carry in the core--in fact, I'm *very* supportive of it (less core stuff to maintain, the better for everyone).  What I want to be careful about is assuming that our typical way of working is *everyone's* way of working.  I find that there's this constant argument in the wild about "most people"--I'm not most people: I type Dvorak, I work in disconnected environments, I have a finely-tuned Vim config that's evolved over years of work, I exercise some of the more interesting corners of Vim, I am a Vim power user. :-)

A package manager would be very nice.  But let's please not go so far as to assume everyone using Neovim will be in a connected environment all the time.  As long as there's a non-painful path to operate is a disconnected environment, I'm definitely supportive of such efforts.

**Comment by jszakmeister** [comment]
> Until now I had not heard anyone speak in favor of continuing to ship the entirety of the Vim runtime files with Neovim. Personally, I strongly prefer compatibility in most cases, I just did not think vimball was such a case.

I'm not sure "compatibility" was ever defined in the sense of what Neovim is going to support out-of-the-box, so it's been left to interpretation.  Perhaps we should nail down what it means a little more.

More than anything, I wanted to point out that vimballs are useful, and I believe they are used.  Plugin managers are great, but they aren't a panacea for those of us running in a disconnected environment.

> I suppose in this case it's a bit of a chicken-and-egg problem :)

What I mean is that there's a path to install vimball support without vimball support already being present. :-)  IOW, we don't necessarily have to maintain it in the core, though I do think it's a useful one to keep around.

**Comment by tarruda** [comment]
>  What I want to be careful about is assuming that our typical way of working is everyone's way of working. I find that there's this constant argument in the wild about "most people"--I'm not most people: I type Dvorak, I work in disconnected environments, I have a finely-tuned Vim config that's evolved over years of work, I exercise some of the more interesting corners of Vim, I am a Vim power user. :-)

Good point. We need to be careful about imposing our own preferences to others, especially because we more people using Neovim :smile:

**Comment by mseeber** [comment]
Maybe a poll could help to gain some insight about user behavior in this case?

Maybe another option would be to provide vimballs functionality as a add on tool/binary but not in the Neovim project itself?

**Comment by justinmk** [comment]
We'll just keep vimball unless/until a better alternative arrives.

**Comment by Pyrohh** [comment]
> We'll just keep vimball unless/until a better alternative arrives.

Sounds good, but does anyone know about `getscript`? I have no clue if it's needed by `vimball`, or even wanted.

**Comment by mtglsk** [comment]
Anyone also has bad indentation in `Local additions:` part (of `:h`)? I was always bugged by this.

Help is not very intuitive at all, quite important doc (sans the contributors part) is `intro.txt`, which is at like line 100. Good luck someone finding it, before opening awful for a newcomer `quickref.txt`.

**Comment by Grimy** [comment]
> wc -l runtime/*/.vim shows 179k lines of vimscript, I wonder how much of that is actually used by the regular vim user?

About 0%. I explicitly remove the default runtime from my runtimepath because it’s a redundant, buggy, inconsistent, unmaintanable and unmaintained mess. I still symlink in `doc` and `autoload` because otherwise some features simply won’t work.

For each language I use, I either install a plugin or write the ftdetect, syntax and indent files myself. Not sure whether this qualifies me as a « regular vim user ».

**Comment by Pyrohh** [comment]
@mtglsk

> Anyone also has bad indentation in Local additions: part (of :h)? I was always bugged by this.

I'll look into that, it shouldn't be too hard to align things properly.

> Help is not very intuitive at all, quite important doc for getting around using the help (sans the contributors part) is intro.txt, which is at like line 100. Good luck someone finding it, before opening awful for a newcomer quickref.txt.

Yes, I definitely agree it's not as concise/clear as I'd like it, but that's for another issue.

@Grimy

> About 0%. I explicitly remove the default runtime from my runtimepath because it’s a redundant, buggy, inconsistent, unmaintanable and unmaintained mess.

This is way too vague, do you have anything in specific you want to mention? I'd like to identify and improve problem spots, but there's just too much docs for one person to go through.

**Comment by Grimy** [comment]
> This is way too vague, do you have anything in specific you want to mention? I'd like to identify and improve problem spots, but there's just too much docs for one person to go through.
I wasn’t talking about the docs, but answering to @tarruda’s remark about VimL files in the default runtime. Sorry for being unclear and off-topic.

**Comment by Pyrohh** [comment]
> I wasn’t talking about the docs, but answering to @tarruda’s remark about VimL files in the default runtime.

Oops, I missed that.

> Sorry for being unclear and off-topic.

It's fine, feel free to open an issue, as I agree `runtime/` could use some refinement.

**Comment by justinmk** [comment]
@Pyrohh getscript/GLVS can be removed. It pulls scripts from vim.sf.net (http://www.vim.org/scripts/). This feature is supported by vundle/neobundle/vim-plug.

So I this issue can be closed, and :+1:  to removing getscript.
